// Phone/Phon.js
import React  ,{ Component } from "react"
import { connect } from "react-redux"
import * as R from 'ramda'

import {fetchPhoneById,addPhoneToBasket} from 'containers/actions/index'
import {getPhoneByid} from 'reducers/selectors'
import s from './phone.module.css'
import { Link } from "react-router-dom"
import BasketCart from 'components/basketCart/basketCart'


class Phone extends Component {
    componentDidMount(){
        this.props.fetchPhoneById(this.props.match.params.id)
    }

    renderFileds(){
        const {phone}=this.props
        const columnField = R.compose(
          R.toPairs,  R.pick([
                'display',
                'camera',
                'music',
                'work',
                'internet',
                'process'
            ])
        )(phone)
        return ( columnField.map(([key,value])=>(
            <div className={s.information} key={key}>
                <div className={s.information_key}>
                    <p>{key}</p>
                </div>
                <div className={s.information_value}>
                    {value}
                </div>
            </div>
        ))
          
        )

    }

    renderContent(){
        const {phone} = this.props
        return (
            
            <div className={s.phone}>
                <div className={s.avatar}>
                    <div>
                        <img className={s.img} src={phone.image}  alt={phone.name} />
                    </div>
                </div>
        <div className={s.info}>{this.renderFileds()}</div>
                <div className ={s.price}>
                    <div className={s.phone_l}>
                    <div className={s.phone_name}>
                       <h1 className={s.h1}> {phone.name}</h1>
                    </div>
        <div className={s.phone_price}>
            <h1 className={s.h1}>${phone.price}</h1></div>
        </div>
                    <div className={s.products}>
                      <h3 className={s.h3}>  Lorem ipsum dolor sit amet consectetur adipisicing elit. Esse architecto voluptatum, aperiam ducimus earum laudantium rerum atque, optio repudiandae maxime eum dolorem similique! Dolores, natus libero nisi dolore at veniam. </h3>
                    </div>
                </div>
            </div>
        )
    }

    renderSidebar(){
        const {phone,addPhoneToBasket}=this.props
        return (
            <div clasName={s.conext}>
               <div className={s.baskets}>
                     <div clasName={s.p}>Quick shop</div>
                <BasketCart/>
               </div> 
                 <div className={s.names1}> 
                     <h1 clasName={s.namePhone}>{phone.name}</h1>
                     <h1 className={s.phone1s}>${phone.price}</h1>
                     
            </div>
        
                <div className={s.links1}>
                <Link to='/'  className={s.link}>Back to store</Link>
                <button
                className={s.btn}
                onClick ={()=>addPhoneToBasket(phone.id)}
                >ADD to Card</button>
                </div>
             
            </div>
        )
    }
        render(){
            const {phone} = this.props

            return (
                <div className={s.container}>
                   
                        <div className={s.content}>
                            {phone && this.renderContent()}
                        </div>
                        <div className={s.sidebar}>
                            { phone && this.renderSidebar()}
                        </div>
                    
                </div>
            )
        }
    }



    const mapStateToProps = state => {
        return{
            phone: getPhoneByid(state, state.phonePage.id)
        }
    }

    const mapDispatchToProps = {
        fetchPhoneById,
        addPhoneToBasket
    }

    export default connect(mapStateToProps, mapDispatchToProps)(Phone)
